stage #2.5/module #3 Spring Core/#5.1. Annotation-based configuration/content.md

```diff
@@ -0,0 +1,470 @@
+behavior, enabling the framework to skip a non-satisfiable injection point through marking it as non-required (i.e., by
+setting the required attribute in `@Autowired` to false):
+
+```
+@Autowired(required = false)
+```
+
+A non-required method will not be called at all if its dependency (or one of its dependencies, in case of multiple
+arguments) is not available. A non-required field will not get populated at all in such cases, leaving its default value
+in place.
+
+Alternatively, you can express the non-required nature of a particular dependency through Java 8's `java.util.Optional`,
+as the following example shows:
+
+```java
+public class MusicRatingService {
+
+    private Optional<MusicRatesFinder> musicRatesFinder;
+
+    @Autowired
+    public void setMusicRatesFinder(Optional<MusicRatesFinder> musicRatesFinder) {
+        this.musicRatesFinder = musicRatesFinder;
+    }
+
+    // ...
+}
+```
+
+As of Spring Framework 5.0, you can also use a `@Nullable annotation` (of any kind in any package - for example,
+`javax.annotation.Nullable` from JSR-305)
+
+```java
+public class MusicRatingService {
+
+    private MusicRatesFinder musicRatesFinder;
+
+    @Autowired
+    public void setMusicRatesFinder(@Nullable MusicRatesFinder musicRatesFinder) {
+        this.musicRatesFinder = musicRatesFinder;
+    }
+
+    // ...
+}
+```
+
+You can also use `@Autowired` for interfaces that are well-known resolvable dependencies: `BeanFactory`,
+`ApplicationContext`, `Environment`, `ResourceLoader`, `ApplicationEventPublisher`, and `MessageSource`. These
+interfaces and their extended interfaces, such as `ConfigurableApplicationContext` or `ResourcePatternResolver`, are
+automatically resolved, with no special setup necessary.
+
+> The `@Autowired`, `@Inject`, `@Value`, and `@Resource` annotations are handled by Spring `BeanPostProcessor`
+> implementations. This means that you cannot apply these annotations within your own `BeanPostProcessor` or
+> `BeanFactoryPostProcessor` types (if any). These types must be 'wired up' explicitly by using XML or a Spring
+> `@Bean` method.
+
+## Annotation-based Autowiring with Qualifiers
+
+When you need more control over the bean candidates selection process, you can use Spring's `@Qualifier` annotation. You
+can associate qualifier values with specific arguments, narrowing the set of type matches so that a specific bean is
+chosen for each argument. In the simplest case, this can be a plain descriptive value, as shown in the following
+example:
+
+```java
+public class MusicRatingService {
+
+    @Autowired
+    @Qualifier("main")
+    private MusicMagazine musicMagazine;
+
+    // ...
+}
+```
+
+You can also specify the `@Qualifier` annotation on individual constructor arguments or method parameters.
+
+The following example shows corresponding bean definitions.
+
+```xml
+<?xml version="1.0" encoding="UTF-8"?>
```

Comment:
I don't think that we really need to give an examples of configuring some settings through xml(annotations alternatives).
Just saying that this is possble and providing one example is enough. And you already have done did it in "IOC Container" chapter

Reply:
This example shows how to define exactly <qualifier> tag in xml config while IOC container chapter shows general beans definition in xml config. Also projects can be different and some projects can still use xml config so I think it would be useful for student to show how to setup somethin in xml as well.

Reply:
Projects can still use xml config - it's really good reason to leave this configuration here